tpl/collections/querify.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 The Hugo Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collections
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/gohugoio/hugo/common/maps"
+	"github.com/spf13/cast"
+)
+
+var (
+	msgExpectedParams string = "expected a map, a slice with an even number of elements, or an even number of scalar values, and each key must be a string"
+	msgZeroLengthKey  string = "one of the keys is an empty string"
+)
+
+// Querify returns a URL query string composed of the given key-value pairs,
+// encoded and sorted by key.
+func (ns *Namespace) Querify(params ...any) (string, error) {
+	if len(params) == 0 {
+		return "", nil
+	}
+
+	if len(params) == 1 {
+		switch v := params[0].(type) {
+		case map[string]any: // created with collections.Dictionary
+			return mapToQueryString(v)
+		case maps.Params: // site configuration or page parameters
+			return mapToQueryString(v)
+		case []string:
+			return stringSliceToQueryString(v)
+		case []any:
+			s, err := interfaceSliceToStringSlice(v)
+			if err != nil {
+				return "", err
+			}
+			return stringSliceToQueryString(s)
+		default:
+			return "", fmt.Errorf("%s", msgExpectedParams)
+		}
+	}
+
+	if len(params)%2 != 0 {
+		return "", fmt.Errorf("%s", msgExpectedParams)
```

Comment:
This is a little bit of a "style" thing, but if the error message never changes, I suggest we replace the `msgExpectedParams` with global error variables (using `errors.New`) and return those.